src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java

```diff
@@ -86,34 +110,100 @@ public MesosCloud(String name, String mesosUrl, String jenkinsUrl)
+  private Optional<MesosAgentSpecTemplate> getSpecForLabel(Label label) {
+    for (MesosAgentSpecTemplate spec : this.mesosAgentSpecTemplates) {
+      if (label.matches(spec.getLabelSet())) {
+        return Optional.of(spec);
+      }
+    }
+    return Optional.empty();
   }
 
   /**
    * Start a Jenkins agent.jar on Mesos.
    *
    * <p>Provide a callback for Jenkins to start a Node.
    *
+   * @param name Name of the Jenkins name and Mesos task.
+   * @param spec The {@link MesosAgentSpecTemplate} that was configured for the Jenkins node.
    * @return A future reference to the launched node.
    */
-  public Future<Node> startAgent() throws Exception {
+  public Future<Node> startAgent(String name, MesosAgentSpecTemplate spec)
+      throws IOException, FormException, URISyntaxException {
     return mesos
-        .enqueueAgent(this, 0.1, 32)
+        .enqueueAgent(this, name, spec)
         .thenCompose(
-            mesosSlave -> {
+            mesosAgent -> {
               try {
-                Jenkins.getInstanceOrNull().addNode(mesosSlave);
-                logger.info("waiting for slave to come online...");
-                return mesosSlave.waitUntilOnlineAsync();
+                Jenkins.get().addNode(mesosAgent);
+                logger.info("waiting for node to come online...");
+                return mesosAgent
+                    .waitUntilOnlineAsync()
+                    .thenApply(
+                        node -> {
+                          logger.info("Agent {} is online", name);
+                          return node;
+                        });
               } catch (Exception ex) {
                 throw new CompletionException(ex);
               }
             })
         .toCompletableFuture();
   }
+
+  @Extension
+  public static class DescriptorImpl extends Descriptor<Cloud> {
+
+    public DescriptorImpl() {
+      load();
+    }
+
+    @Override
+    public String getDisplayName() {
+      return "Mesos Cloud";
+    }
+
+    // TODO: validate URLs
+
+    /** Test connection from configuration page. */
+    public FormValidation doTestConnection(
+        @QueryParameter("mesosMasterUrl") String mesosMasterUrl) {
+      throw new NotImplementedException("Connection testing is not supported yet.");
+    }
+  }
+
+  // Getters
+
+  public String getMesosMasterUrl() {
+    return this.mesosMasterUrl.toString();
+  }
+
+  public String getFrameworkName() {
+    return this.mesos.getFrameworkName();
+  }
+
+  public String getJenkinsUrl() {
+    return this.jenkinsUrl.toString();
+  }
+
+  public String getAgentUser() {
+    return "kjeschkies";
+  }
+
+  public String getRole() {
+    return "*";
+  }
+
+  /** @return Number of launching agents that are not connected yet. */
+  public synchronized int getPending() {
+    return toIntExact(mesos.getState().values().stream().filter(MesosAgent::isPending).count());
+  }
+
+  public MesosApi getMesosClient() {
+    return this.mesos;
```

Comment:
can we rename this? the type  is MesosApi the variable is mesos and the method is getMesosClient which gets confusing

Reply:
Yea, that is odd indeed.